### PR TITLE
SETI-1706: Check that SHA of PR head fetched by merger is correct

### PR DESCRIPTION
**Pull request by janpriessnitz**
git ref pull/X/head is sometimes wrong (error of Github).
Catch these errors by checking that SHA of commit fetched by merger
equals SHA of PR head acquired from Github webhook event JSON.

### CODE REVIEW

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/130/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/159/ : SUCCESS in 7m 23s

File: zuul/connection/github.py [hunk 1]
```diff
@@ -203,6 +203,7 @@ def _pull_request_to_event(self, pr_body):
         event.updated_at = pr_body.get('updated_at')
         event.branch = base.get('ref')
         event.refspec = "refs/pull/" + str(pr_body.get('number')) + "/head"
+        event.refspec_sha = head.get('sha')
         event.patch_number = head.get('sha')
```
Comment (hrubi) [inline zuul/connection/github.py]: You could've reused the `event.patch_number`. Is there a reason you didn't want to? Like that it serves different purpose?

File: zuul/exceptions.py [hunk 2]
```diff
@@ -37,3 +37,10 @@ class MergeFailure(Exception):
 
 class HeadBranchModified(Exception):
     pass
+
+
+class ExpectedSHAMismatch(Exception):
+    def __init__(self):
+        self.message = "Merger did not fetch expected head commit " \
+                       "(fetched a different one)."
+        super(ExpectedSHAMismatch, self).__init__(self.message)
```
Comment (hrubi) [inline zuul/exceptions.py]: Cosmetics - missing newline at the end.

File: zuul/connection/github.py (hunk 1 above)
Comment (janpriessnitz) [reply zuul/connection/github.py]: In github connection it's the same, so it won't be any difference. But in gerrit connection the two variables are different. I know we do not use the SHA check in gerrit pipelines, so technically, there's no need for a separate variable. But if we decided in the future to implement it on gerrit pipelines, one variable wouldn't suffice. Plus the code would be even more messy if some future contributor tried to understand it. What do you think?

**Comment by hrubi** [comment]
yeah, I agree with you. keep it as is.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/131/ : SUCCESS in 10s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/160/ : SUCCESS in 6m 53s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/132/ : SUCCESS in 9s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/161/ : SUCCESS in 9m 18s

File: tests/test_github.py [hunk 3]
```diff
@@ -641,3 +641,24 @@ def test_head_branch_modified(self):
         self.assertIn("Error merging pull request:"
                       " Head branch was modified.", A.comments)
         self.assertEqual(False, A.is_merged)
+
+    def test_head_commit_sha_mismatch(self):
+        """Test that zuul fails when trying to merge different head commit"""
```
Comment (michalvanco) [inline tests/test_github.py]: multiline comment

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/133/ : SUCCESS in 8s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/162/ : FAILURE in 7m 36s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/zuul-opensource-repository-check/134/ : SUCCESS in 9s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/163/ : SUCCESS in 16m 51s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://ci.intgdc.com/job/zuul-merge-compliance-check/28/ : SUCCESS in 16s
- https://ci.intgdc.com/job/zuul-tox-zuul-el7/164/ : SUCCESS in 6m 40s

**Pull request merged**